### PR TITLE
Add Apple log sink.

### PR DESCRIPTION
**Pull request by edgchen1**
**Description**
Add a log sink for Apple platforms. This version uses NSLog().

**Motivation and Context**
Enable logging with the Apple logging system.

### CODE REVIEW

File: cmake/onnxruntime_providers.cmake [hunk 1]
```diff
@@ -669,6 +669,10 @@ if (onnxruntime_USE_COREML)
     message(FATAL_ERROR "CoreML EP can not be used in a basic minimal build. Please build with '--minimal_build extended'")
   endif()
 
+  if(NOT APPLE)
+    message(FATAL_ERROR "CoreML EP must be built on an Apple platform.")
+  endif()
+
   add_compile_definitions(USE_COREML=1)
```
Comment (skottmckay) [inline cmake/onnxruntime_providers.cmake:676]: Not sure this is correct. We need to support a build on non-apple platforms in order to create an ORT format model that's aware of what the CoreML EP can handle. 

This build only supports GetCapability, and the Compile returns an error (e.g. [https://github.com/microsoft/onnxruntime/blob/cdd2129fed4c7583011338485143e641084366a0/onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc#L459-L474](https://github.com/microsoft/onnxruntime/blob/cdd2129fed4c7583011338485143e641084366a0/onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc#L459-L474))

Comment (skottmckay) [reply cmake/onnxruntime_providers.cmake:676]: At least until we implement runtime replay of optimizations, as we wouldn't need to create an ORT format model that's aware of what nodes NNAPI or CoreML could take.

Comment (edgchen1) [reply cmake/onnxruntime_providers.cmake:676]: The CoreML EP has some Objective-C++ files. Does that mean we support building those on non-Apple platforms?

Comment (gwang-msft) [reply cmake/onnxruntime_providers.cmake:676]: The GetCapability of CoreML EP is not using any objc code or any apple platform library.
But for now the CoreML EP, GetCapability and Compile are not yet separated, so this change is ok. 
We can separate those later when we add coreml to the conversion script. (need to add a bunch of #ifdef and change the cmake files)

File: onnxruntime/core/session/ort_env.cc [hunk 2]
```diff
@@ -55,20 +57,22 @@ OrtEnv* OrtEnv::GetInstance(const OrtEnv::LoggingManagerConstructionInfo& lm_inf
...
+                                                                       lm_info.logger_param);
       lmgr.reset(new LoggingManager(std::move(logger),
                                     static_cast<Severity>(lm_info.default_warning_level),
                                     false,
                                     LoggingManager::InstanceType::Default,
                                     &name));
     } else {
-#ifdef __ANDROID__
-      ISink* sink = new AndroidLogSink();
+#if defined(__ANDROID__)
```
Comment (gwang-msft) [inline onnxruntime/core/session/ort_env.cc:67]: Do we want to add an separated function to host all these #ifdef's?

File: onnxruntime/core/providers/coreml/model/model.mm [hunk 3]
```diff
@@ -167,6 +167,7 @@ - (void)cleanup {
 
 - (void)dealloc {
   [self cleanup];
+  [super dealloc];
```
[the remark this answers is not in the document]
Comment (edgchen1) [reply onnxruntime/core/providers/coreml/model/model.mm:170]: -fobjc-arc is somehow not going through in my tests. trying to figure that out.

File: cmake/onnxruntime_providers.cmake (hunk 1 above)
Comment (edgchen1) [reply cmake/onnxruntime_providers.cmake]: handling this another way now, removed the if(APPLE) check

**Review by skottmckay — APPROVED** [review]
:shipit:

**Pull request merged**